clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +375,324 @@ impl FunctionType {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
+            )?;
+        }
+        Ok(returns.clone())
+    }
+}
+
+pub fn trait_check_trait_compliance<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_trait_identifier: &TraitIdentifier,
+    actual_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    expected_trait_identifier: &TraitIdentifier,
+    expected_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> CheckResult<()> {
+    // Shortcut for the simple case when the two traits are the same.
+    if actual_trait_identifier == expected_trait_identifier {
+        return Ok(());
+    }
+
+    for (func_name, expected_sig) in expected_trait.iter() {
+        if let Some(func) = actual_trait.get(func_name) {
+            let args_iter = expected_sig.args.iter().zip(func.args.iter());
+            for (expected_type, actual_type) in args_iter {
+                if inner_type_check_type(
+                    db,
+                    contract_context,
+                    actual_type,
+                    expected_type,
+                    clarity_version,
+                    1,
+                    tracker,
+                )
+                .is_err()
+                {
+                    return Err(CheckErrors::IncompatibleTrait(
+                        expected_trait_identifier.clone(),
+                        actual_trait_identifier.clone(),
+                    )
+                    .into());
                 }
-                (expected_type, value) => {
-                    if !expected_type.admits(&value) {
-                        let actual_type = TypeSignature::type_of(&value);
-                        return Err(
-                            CheckErrors::TypeError(expected_type.clone(), actual_type).into()
-                        );
+            }
+            if inner_type_check_type(
+                db,
+                contract_context,
+                &func.returns,
+                &expected_sig.returns,
+                clarity_version,
+                1,
+                tracker,
+            )
+            .is_err()
+            {
+                return Err(CheckErrors::IncompatibleTrait(
+                    expected_trait_identifier.clone(),
+                    actual_trait_identifier.clone(),
+                )
+                .into());
+            }
+        } else {
+            return Err(CheckErrors::IncompatibleTrait(
+                expected_trait_identifier.clone(),
+                actual_trait_identifier.clone(),
+            )
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn inner_type_check_type<T: CostTracker>(
```

Comment:
Can you add a rustdoc to this method that describes at a high level what this method is trying to do. Like:

```rust
/// Check if `expected_type` admits `actual_type`, handling traits and callable types
/// through invoking trait compliance checks.
```

Reply:
Added in a4e6ed128